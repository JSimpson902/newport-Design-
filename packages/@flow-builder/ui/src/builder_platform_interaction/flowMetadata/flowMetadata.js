export const FLOW_PROCESS_TYPE = {
    AUTO_LAUNCHED_FLOW : 'AutoLaunchedFlow',
    FLOW : 'Flow'
};

export const ELEMENT_TYPE = {
    ACTION_CALL: 'ACTION_CALL',
    ASSIGNMENT: 'ASSIGNMENT',
    VARIABLE: 'VARIABLE',
    DECISION: 'DECISION',
    DECISION_WITH_MODIFIED_AND_DELETED_OUTCOMES: 'DECISION_WITH_MODIFIED_AND_DELETED_OUTCOMES',
    APEX_PLUGIN_CALL: 'APEX_PLUGIN_CALL',
    APEX_CALL: 'APEX_CALL',
    EMAIL_ALERT: 'EMAIL_ALERT',
    CHOICE: 'CHOICE',
    CONSTANT: 'CONSTANT',
    DYNAMIC_CHOICE_SET: 'DYNAMIC_CHOICE_SET',
    FORMULA: 'FORMULA',
    STAGE: 'STAGE',
    TEXT_TEMPLATE: 'TEXT_TEMPLATE',
    LOOP: 'LOOP',
    OUTCOME: 'OUTCOME',
    RECORD_CREATE: 'RECORD_CREATE',
    RECORD_DELETE: 'RECORD_DELETE',
    RECORD_LOOKUP: 'RECORD_LOOKUP',
    RECORD_UPDATE: 'RECORD_UPDATE',
    SCREEN: 'SCREEN',
    STEP: 'STEP',
    SUBFLOW: 'SUBFLOW',
    START_ELEMENT: 'START_ELEMENT',
    WAIT: 'WAIT',
    FLOW_PROPERTIES: 'FLOW_PROPERTIES',
    DEFAULT: 'defaultElement'
};

export const CONNECTOR_TYPE = {
    REGULAR: 'REGULAR',
    FAULT: 'FAULT',
    DEFAULT: 'DEFAULT',
    START: 'START',
    LOOP_NEXT: 'LOOP_NEXT',
    LOOP_END: 'LOOP_END'
};

export const SUB_ELEMENT_TYPE = {
    ASSIGNMENT_ITEM: 'ASSIGNMENT_ITEM',
    CONDITION: 'CONDITION',
    RECORD_LOOKUP_FILTER_ITEM: 'RECORD_LOOKUP_FILTER_ITEM',
    RECORD_LOOKUP_FIELD: 'RECORD_LOOKUP_FIELD',
};

export const ACTION_TYPE = {
    APEX: 'apex',
    EMAIL_ALERT: 'emailAlert',
    QUICK_ACTION: 'quickAction',
    FLOW: 'flow',
    COMPONENT: 'component'
};

export const METADATA_KEY = {
    ACTION_CALLS: 'actionCalls',
    APEX_PLUGIN_CALLS: 'apexPluginCalls',
    ASSIGNMENTS: 'assignments',
    DECISIONS: 'decisions',
    WAITS: 'waits',
    FORMULAS: 'formulas',
    VARIABLES: 'variables',
    CONSTANTS: 'constants',
    SUBFLOWS: 'subflows',
    RECORD_CREATE: 'recordCreates',
    RECORD_LOOKUP: 'recordLookups',
    RECORD_DELETE: 'recordDeletes',
    RECORD_UPDATE: 'recordUpdates',
    LOOPS: 'loops',
    SCREENS: 'screens'
};

/**
 * AND and OR are flow values which are used by the backend as well as being as used
 * on the frontend (combobox values).  CUSTOM_LOGIC is used only by the frontend as a
 * combobox value
 * @type {{AND: string, OR: string, ALWAYS: string, CUSTOM_LOGIC: string}}
 */
export const CONDITION_LOGIC = {
    AND: 'and',
    OR: 'or',
    CUSTOM_LOGIC: 'custom_logic',
    NO_CONDITIONS: 'no_conditions'
};

/**
 * The top level process metadata values to save with each flow. This includes the BuilderType
 * which helps us identify flow versions created by the Lightning Flow Builder.
 */
export const PROCESS_METADATA_VALUES = [
    {
        name: 'BuilderType',
        value: {
            stringValue: 'LightningFlowBuilder'
        }
    }
];

/**
 * List of Flow metadata fields that can be references to other Flow elements
 * i.e. fields whose values need to be converted from dev names to GUIDs when loading the flow into the UI client and vice versa.
 */
export const REFERENCE_FIELDS = new Set([
    'elementReference',
    'targetReference',
    'assignToReference',
    'assignNextValueToReference',
    'collectionReference',
    'inputReference',
    'outputReference',
    'assignRecordIdToReference',
    'leftValueReference',
    'outcomeReference',
    'childReference',
    'leftHandSide',
    'defaultValue',
    'source',
    'target',
    'rightHandSideGuid'
]);

/**
 * List of Flow metadata fields that can contain {!reference} tags referencing flow elements
 * These would also need to be replaced
 */
export const TEMPLATE_FIELDS = new Set([
    // TODO fill out more values and confirm that all FEROV stringValue fields allow this behaviour
    'stringValue', // field of a ferov
    'rightHandSide', // field of a ferov
    'expression',  // represents body of a formula
    'text',        // represents body of a text template
    'fieldText',   // body of screen field of type Display Text
    'helpText',     // help text for a screen or screen field
    'interviewLabel' // interview label for the flow properties
]);

export const EXPRESSION_RE = /\{!([^}]+)\}/g;
