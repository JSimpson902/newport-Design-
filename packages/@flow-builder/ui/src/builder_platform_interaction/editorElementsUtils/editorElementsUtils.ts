// @ts-nocheck
import { getConfigForElement, getConfigForElementType } from 'builder_platform_interaction/elementConfig';
import { generateGuid } from 'builder_platform_interaction/storeLib';
import { ELEMENT_TYPE, ACTION_TYPE, ACTION_TYPE_TO_ELEMENT_TYPE } from 'builder_platform_interaction/flowMetadata';
import { loggingUtils } from 'builder_platform_interaction/sharedUtils';
import { getProcessType } from 'builder_platform_interaction/storeUtils';
import { isOrchestrator } from 'builder_platform_interaction/processTypeLib';

const { logMetricsServiceErrorTransaction } = loggingUtils;
/**
 * Transforms elements into a form that is usable by lightning-tree-grid. These
 * are grouped by element category so that they can more easily be placed into
 * sections.
 *
 * @param {Object}
 *            elements list of all the elements
 * @param elements
 * @param palette
 * @returns {Object} a mapping of element type to a list of
 *          lightning-tree-grid-items
 */
const mutateElements = (elements, palette) =>
    palette.headers.reduce(
        (acc, { headerLabel, headerItems, headerFreeformVisibility, headerAutolayoutVisibility }) => {
            try {
                if (headerLabel && headerItems) {
                    headerItems.forEach((headerItem) => {
                        const filteredElements = elements.filter((el) => {
                            if (
                                headerItem.type === 'element' ||
                                (headerItem.type === 'shortcut' && headerItem.shortcutEnumType === 'element')
                            ) {
                                return (
                                    headerItem.name === el.elementType || headerItem.shortcutEnumName === el.elementType
                                );
                            }
                            if (
                                headerItem.type === 'elementSubtype' ||
                                (headerItem.type === 'shortcut' && headerItem.shortcutEnumType === 'elementSubtype')
                            ) {
                                return headerItem.name === el.name || headerItem.shortcutEnumName === el.name;
                            }

                            return false;
                        });

                        if (
                            headerItem.type === 'action' ||
                            (headerItem.type === 'shortcut' && headerItem.shortcutEnumType === 'action')
                        ) {
                            const item = {
                                elementType: ELEMENT_TYPE.ACTION_CALL,
                                actionLabel: headerItem.actionLabel ?? headerItem.shortcutLabel,
                                actionType: headerItem.name ?? headerItem.shortcutEnumName,
                                actionName: headerItem.name ?? headerItem.shortcutEnumName,
                                actionIsStandard: headerItem.actionIsStandard,
                                description: headerItem.actionDescription ?? headerItem.shortcutDescription
                            };
                            filteredElements.push(item);
                        }

                        if (filteredElements.length > 0) {
                            filteredElements.forEach((element) => {
                                const elementType = element.elementType;
                                const elementSubtype = element.isElementSubtype ? element.name : null;
                                const getElementTypeByActionType =
                                    ACTION_TYPE_TO_ELEMENT_TYPE[headerItem.shortcutEnumName];
                                const configDetails =
                                    element.actionType && !element.actionIsStandard && getElementTypeByActionType
                                        ? getConfigForElementType(getElementTypeByActionType)
                                        : getConfigForElement(element);
                                const { nodeConfig, labels, canHaveFaultConnector } = configDetails;
                                const label =
                                    element.actionLabel || headerItem.shortcutLabel || (labels && labels.leftPanel);
                                const actionType = element.actionType;
                                const actionName = element.actionName;
                                const actionIsStandard = element.actionIsStandard;
                                const {
                                    iconName,
                                    dragImageSrc,
                                    iconBackgroundColor,
                                    iconShape,
                                    iconSize,
                                    dynamicNodeComponent,
                                    dynamicNodeComponentSelector
                                } = nodeConfig;

                                // Favor using the description supplied by the element if it exists, otherwise
                                // fall back to the default description.
                                const description =
                                    element.description ||
                                    headerItem.shortcutDescription ||
                                    (isOrchestrator(getProcessType())
                                        ? nodeConfig.orchestratorDescription || nodeConfig.description
                                        : nodeConfig.description);

                                if (!acc[headerLabel]) {
                                    acc[headerLabel] = [];
                                }
                                const item = {
                                    guid: generateGuid(),
                                    iconName,
                                    dragImageSrc,
                                    iconBackgroundColor,
                                    label,
                                    description,
                                    elementType,
                                    elementSubtype,
                                    actionType,
                                    actionName,
                                    actionIsStandard,
                                    canHaveFaultConnector,
                                    iconShape,
                                    iconSize,
                                    dynamicNodeComponent,
                                    dynamicNodeComponentSelector,
                                    freeformVisible: headerFreeformVisibility,
                                    alcVisible: headerAutolayoutVisibility
                                };
                                acc[headerLabel].push(item);
                            });
                        }
                    });
                }
            } catch (e) {
                // if an element is invalid, just log it but don't stop the rest from loading
                logMetricsServiceErrorTransaction(JSON.stringify(e, Object.getOwnPropertyNames(e)));
            }
            return acc;
        },
        {}
    );

/**
 * Combines elements into their respective groupings in a form that is usable by
 * lightning-tree-grid.
 *
 * @param {Object}
 *            elements list of all the elements
 * @param elements
 * @param palette
 * @returns {Array} collection of lightning-tree-grid items
 */
export const getElementSections = (elements, palette) => {
    if (!elements || Object.keys(elements).length === 0) {
        return [];
    }

    const elementMap = mutateElements(elements, palette);
    const elementSections = Object.keys(elementMap).reduce((acc, name) => {
        const section = {};
        section._children = elementMap[name];
        section.guid = generateGuid();
        section.label = name;
        acc.push(section);
        return acc;
    }, []);

    return elementSections;
};
