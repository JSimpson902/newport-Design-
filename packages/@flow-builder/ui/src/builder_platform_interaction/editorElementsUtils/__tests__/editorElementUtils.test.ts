// @ts-nocheck
import { getElementSections } from 'builder_platform_interaction/editorElementsUtils';
import { ELEMENT_TYPE } from 'builder_platform_interaction/flowMetadata';

jest.mock('builder_platform_interaction/storeLib', () => {
    const actual = jest.requireActual('builder_platform_interaction/storeLib');
    return {
        deepCopy: actual.deepCopy,
        generateGuid: jest.fn().mockImplementation(() => {
            return 'testGUID';
        })
    };
});

describe('element-lib', () => {
    describe('When no elements are available', () => {
        it('returns an empty list when elements is undefined', () => {
            expect(getElementSections()).toEqual([]);
        });

        it('returns an empty list when elements is null', () => {
            expect(getElementSections(null)).toEqual([]);
        });

        it('returns an empty list when elements is empty', () => {
            expect(getElementSections({})).toEqual([]);
        });
    });

    describe('When elements are available', () => {
        it('returns the expected sections list based on palette', () => {
            const unsortedElements = [
                { elementType: ELEMENT_TYPE.WAIT },
                { elementType: ELEMENT_TYPE.ASSIGNMENT },
                { elementType: ELEMENT_TYPE.SCREEN },
                { elementType: ELEMENT_TYPE.RECORD_LOOKUP },
                { elementType: ELEMENT_TYPE.RECORD_CREATE },
                { elementType: ELEMENT_TYPE.ACTION_CALL },
                { type: 'marketingEmail', name: 'marketingEmail', label: 'testActionLabel' }
            ];

            const palette = {
                headers: [
                    {
                        headerLabel: 'FlowBuilderLeftPanelElements.flowInteractionComponentsLabel',
                        headerItems: [
                            { type: 'element', name: ELEMENT_TYPE.SCREEN },
                            { type: 'element', name: ELEMENT_TYPE.ACTION_CALL },
                            { type: 'element', name: ELEMENT_TYPE.LOOP }
                        ]
                    },
                    {
                        headerLabel: 'FlowBuilderLeftPanelElements.flowControlLogicLabel',
                        headerItems: [
                            { type: 'element', name: ELEMENT_TYPE.ASSIGNMENT },
                            { type: 'element', name: ELEMENT_TYPE.WAIT },
                            { type: 'action', name: 'marketingEmail' }
                        ]
                    },
                    {
                        headerLabel: 'FlowBuilderLeftPanelElements.flowControlDataOperationsLabel',
                        headerItems: [
                            { type: 'element', name: ELEMENT_TYPE.RECORD_CREATE },
                            { type: 'element', name: ELEMENT_TYPE.RECORD_LOOKUP }
                        ]
                    }
                ]
            };

            const expectedElementSections = [
                {
                    _children: [{ elementType: ELEMENT_TYPE.SCREEN }, { elementType: ELEMENT_TYPE.ACTION_CALL }],
                    guid: 'testGUID',
                    label: 'FlowBuilderLeftPanelElements.flowInteractionComponentsLabel'
                },
                {
                    _children: [
                        { elementType: ELEMENT_TYPE.ASSIGNMENT },
                        { elementType: ELEMENT_TYPE.WAIT },
                        {
                            elementType: ELEMENT_TYPE.ACTION_CALL,
                            actionType: 'marketingEmail',
                            actionName: 'marketingEmail'
                        }
                    ],
                    guid: 'testGUID',
                    label: 'FlowBuilderLeftPanelElements.flowControlLogicLabel'
                },
                {
                    _children: [
                        { elementType: ELEMENT_TYPE.RECORD_CREATE },
                        { elementType: ELEMENT_TYPE.RECORD_LOOKUP }
                    ],
                    guid: 'testGUID',
                    label: 'FlowBuilderLeftPanelElements.flowControlDataOperationsLabel'
                }
            ];

            expect(getElementSections(unsortedElements, palette)).toMatchObject(expectedElementSections);
        });
    });
});
