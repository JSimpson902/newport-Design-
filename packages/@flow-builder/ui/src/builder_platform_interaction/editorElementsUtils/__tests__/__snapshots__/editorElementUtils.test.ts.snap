// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`element-lib When elements are available returns the expected sections list based on palette 1`] = `
Array [
  Object {
    "_children": Array [
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "QuickLinksLabelAndDescription.recordUpdateDescription",
        "dragImageSrc": "/flow/icons/large/record_update.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "RecordUpdate",
        "freeformVisible": false,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:record_update",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "QuickLinksLabelAndDescription.recordUpdateLabel",
        "supportsBranching": undefined,
      },
      Object {
        "actionIconName": undefined,
        "actionIsStandard": false,
        "actionName": "emailAlert",
        "actionType": "emailAlert",
        "canHaveFaultConnector": undefined,
        "description": "QuickLinksLabelAndDescription.emailAlertDescription",
        "dragImageSrc": "/flow/icons/large/email_alert.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "ActionCall",
        "freeformVisible": false,
        "guid": "testGUID",
        "iconBackgroundColor": "background-navy",
        "iconName": "standard:email",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "QuickLinksLabelAndDescription.emailAlertLabel",
        "supportsBranching": undefined,
      },
    ],
    "guid": "testGUID",
    "label": "FlowBuilderLeftPanelElements.quickLinksLabel",
  },
  Object {
    "_children": Array [
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.screenComponentDescription",
        "dragImageSrc": "/flow/icons/large/screen.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "Screen",
        "freeformVisible": true,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:screen",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.screenComponentLabel",
        "supportsBranching": undefined,
      },
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.actionFlowComponentDescription",
        "dragImageSrc": "/flow/icons/large/action_call.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "ActionCall",
        "freeformVisible": true,
        "guid": "testGUID",
        "iconBackgroundColor": "background-navy",
        "iconName": "standard:custom_notification",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.actionFlowComponentLabel",
        "supportsBranching": undefined,
      },
    ],
    "guid": "testGUID",
    "label": "FlowBuilderLeftPanelElements.flowInteractionComponentsLabel",
  },
  Object {
    "_children": Array [
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.assignmentLogicDescription",
        "dragImageSrc": "/flow/icons/large/assignment.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "Assignment",
        "freeformVisible": true,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:assignment",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.assignmentLogicLabel",
        "supportsBranching": undefined,
      },
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.waitLogicDescription",
        "dragImageSrc": "/flow/icons/large/wait.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "Wait",
        "freeformVisible": true,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:waits",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.waitLogicLabel",
        "supportsBranching": undefined,
      },
      Object {
        "actionIconName": undefined,
        "actionIsStandard": true,
        "actionName": "marketingEmail",
        "actionType": "marketingEmail",
        "canHaveFaultConnector": undefined,
        "description": "Send an email where you specify the subject, body, and recipients",
        "dragImageSrc": "/flow/icons/large/action_call.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "ActionCall",
        "freeformVisible": true,
        "guid": "testGUID",
        "iconBackgroundColor": "background-navy",
        "iconName": "standard:custom_notification",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "Marketing E-Mail",
        "supportsBranching": undefined,
      },
    ],
    "guid": "testGUID",
    "label": "FlowBuilderLeftPanelElements.flowControlLogicLabel",
  },
  Object {
    "_children": Array [
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.createDataOperationDescription",
        "dragImageSrc": "/flow/icons/large/record_create.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "RecordCreate",
        "freeformVisible": undefined,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:record_create",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.createDataOperationLabel",
        "supportsBranching": undefined,
      },
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.updateDataOperationDescription",
        "dragImageSrc": "/flow/icons/large/record_update.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "RecordUpdate",
        "freeformVisible": undefined,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:record_update",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.updateDataOperationLabel",
        "supportsBranching": undefined,
      },
      Object {
        "actionIconName": undefined,
        "actionIsStandard": undefined,
        "actionName": undefined,
        "actionType": undefined,
        "canHaveFaultConnector": undefined,
        "description": "FlowBuilderLeftPanelElements.lookupDataOperationDescription",
        "dragImageSrc": "/flow/icons/large/record_lookup.png",
        "dynamicNodeComponent": undefined,
        "dynamicNodeComponentSelector": undefined,
        "elementSubtype": null,
        "elementType": "RecordQuery",
        "freeformVisible": undefined,
        "guid": "testGUID",
        "iconBackgroundColor": undefined,
        "iconName": "standard:record_lookup",
        "iconShape": undefined,
        "iconSize": undefined,
        "label": "FlowBuilderLeftPanelElements.lookupDataOperationLabel",
        "supportsBranching": undefined,
      },
    ],
    "guid": "testGUID",
    "label": "FlowBuilderLeftPanelElements.flowControlDataOperationsLabel",
  },
]
`;
